lib/elixir/lib/calendar.ex

```diff
@@ -789,28 +945,93 @@ defmodule Time do
       _ -> :eq
     end
   end
 
+  @doc """
+  Converts the Time struct to a different calendar.
+  """
+  @spec convert(Time.t, Calendar.calendar) :: Time.t
+  # No conversion required if already of the target calendar.
+  def convert(%Time{calendar: calendar} = time, calendar) do
+    {:ok, time}
+  end
+
+  def convert(%Time{} = time, calendar) do
+    result_time =
+      time
+      |> to_day_fraction
+      |> calendar.time_from_day_fraction
+    {:ok, result_time}
+  end
+
+  @spec convert!(Time.t, Calendar.calendar) :: Time.t
+  def convert!(time, calendar) do
+    case convert(time, calendar) do
+      {:ok, value} ->
+        value
+      {:error, reason} ->
+        raise ArgumentError, "cannot convert #{inspect time} to target calendar #{calendar}, reason: #{inspect reason}"
+    end
+  end
+
+  @doc """
+  Returns a day_fraction that represents the difference
+  between two Time Structs.
+  """
+  @spec diff(Time.t, Time.t) :: Calendar.rata_die
+  def diff(%Time{} = time1, %Time{} = time2) do
+    {parts1, ppd1} = to_day_fraction(time1)
+    {parts2, ppd2} = to_day_fraction(time2)
+
+    diff_ppd = ppd1 * ppd2
+    diff_parts = parts1 * ppd2 - parts2 * ppd1
+
+    # Keep integers in day fraction low.
+    gcd = gcd(diff_parts, diff_ppd)
+    diff_parts = div(diff_parts, gcd)
+    diff_ppd = div(diff_ppd, gcd)
+
+    {diff_parts, diff_ppd}
+  end
+
   ## Helpers
 
   defp to_tuple(%{hour: hour, minute: minute, second: second, microsecond: {microsecond, _precision}}) do
     {hour, minute, second, microsecond}
   end
 
+  defp to_day_fraction(%{hour: hour, minute: minute, second: second, microsecond: {_, _} = microsecond, calendar: calendar}) do
+    calendar.time_to_day_fraction(hour, minute, second, microsecond)
+  end
+
+  defp to_day_fraction(%{hour: hour, minute: minute, second: second, microsecond: microsecond, calendar: calendar}) do
+    calendar.time_to_day_fraction(hour, minute, second, {microsecond, 0})
+  end
+
+  @spec gcd(integer, integer) :: non_neg_integer
+  defp gcd(a, 0), do: abs(a)
+
+  defp gcd(0, b), do: abs(b)
+  defp gcd(a, b) when a < 0 or b < 0, do: gcd(abs(a), abs(b))
+  defp gcd(a, b), do: gcd(b, rem(a,b))
+
   defimpl String.Chars do
-    def to_string(%{hour: hour, minute: minute, second: second, microsecond: microsecond}) do
-      Calendar.ISO.time_to_string(hour, minute, second, microsecond)
+    def to_string(%{hour: hour, minute: minute, second: second, microsecond: microsecond, calendar: calendar}) do
+      calendar.time_to_string(hour, minute, second, microsecond)
     end
   end
 
   defimpl Inspect do
-    def inspect(%{hour: hour, minute: minute, second: second, microsecond: microsecond}, _) do
-      "~T[" <> Calendar.ISO.time_to_string(hour, minute, second, microsecond) <> "]"
+    def inspect(%{hour: hour, minute: minute, second: second, microsecond: microsecond, calendar: calendar}, _) do
+      "~T[" <> calendar.time_to_string(hour, minute, second, microsecond) <> "]"
```

Comment:
The sigil only works on ISO. So we need to explicitly match on calendar: Calendar.ISO and then fallback to full representation otherwise.

See how it is done on Date or NaiveDateTime.